[Shui Jie] ip

sj1999-BIT opened the pull request:
### **Ultimate Normal Scheduler**

> I am not more gifted than anybody else. I am just more curious than the average person and I will not give up a problem until I have found the proper solution.
> -Albert Einstein


The Ultimate Normal Scheduler has one aim and **one aim** only, to make support you to be the ultimate Task Master!

- ```text-based cause the GUI has not been done yet.```


1. download it from [here](https://www.youtube.com/watch?v=2vi_KMESJ7o).
2. double click it.
3. add your tasks, remove them, whatever you want!
4. let it remember your tasks for you. That's all it's good for! 😉 
5. And it is FREE! 

Features:

- [ ]  Managing tasks like a nagging mother
- [ ]  Managing deadlines (coming soon once the engineers stop throwing bananas)
- [ ]  Reminders (coming soon to computers near you)

Add this into the code to run it!
`public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}`

Commit:
level0

Added intro and bye

Commit:
level-1

Added the ability to echo

Commit:
Level 2. Add, List

Used an ArrayList to store list of items

Commit:
Level 3. Mark as Done

Simplified overall structure. Added an ArrayList as global variable. changed the property of done.

Commit:
Level 4. ToDos, Events, Deadlines

Added  documentation to all the functions. Changed the choiceOfAction function such that it can recognize different task.

setDate function to create the date String

Commit:
A-Inheritance

Changed the structure of the program such that the three task types are all subtask of the main Task

Commit:
A-TextUiTesting

Created a test inputs and outputs

Commit:
Level 5. Handle Errors and A-Exceptions

handle errors involving format problems for all three tasks. Exceptions is utilized.

Commit:
Level 6. Delete

Implement a deleteTask function to remove task when needed

Commit:
shui jie test

Commit:
Revert "shui jie test"

This reverts commit af273dc0070530c210fbd930db578e4b74a911eb.

Commit:
Level 7. Save

Added 3 functions to save list into file, and then read them upon reopening.

Commit:
Level 8. Dates and Times

Added in LocalDate and DateTimeFormatter.

Changed Deadline and Events Task restrictors as the format has now changed.

ambroseboo commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,447 @@
+
 public class Duke {
+
+    private String line = "__________________________________\n";
```
I feel this should be a final field?

KelvinSoo replied:
It seems that line is a constant. Maybe calling it LINE is more suitable for the coding standard.

ambroseboo commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,447 @@
+         * get isDone
+         * @return
+         */
+        public boolean getIsDone() {
```
Perhaps this should just be isDone()?

ambroseboo commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,447 @@
+            return output + " )";
+        }
+        /**
+         * modify the task message if done.
```
I think you should try to capitalise all JavaDocs descriptions.

ambroseboo commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,447 @@
+     * @param input
+     * @return boolean
+     */
+    private boolean choiceOfAction(String input) {
```
Since this is a function that returns a boolean result, I do feel the naming should be isSetDone or isNewTask.

ambroseboo commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,447 @@
+                writer.write(formatTaskIntoCommands(task));
+            }
+            writer.close();
+        }catch (IOException io) {
```
Spacing error.

ambroseboo commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,447 @@
+                try {
+                    Task newTask = createTask(input);
+                    addTask(newTask);
+                } catch (TodoException tx) {
```
I feel that seperating exception for each class is quite unneccesary. Simply a TaskException would be easier to understand.

KelvinSoo commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,447 @@
+     */
+    private void greet() {
+            System.out.println(line);
+            String logo = " ____        _        \n"
```
Would it be better for logo to be declared as a constant?

KelvinSoo commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,447 @@
+                }
+            }
+        }catch (FileNotFoundException f) {
+        }catch(IOException io){
```
There should be a space before the open curly bracket.

KelvinSoo commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,447 @@
+         */
+        public TodoTasks(String action) throws TodoException{
+            super(action, "[T][] ", TaskTypes.Todo);
+            if (action.split(" ").length < 2){
```
There should be a space before the curly bracket.

KelvinSoo reviewed (commented):
Overall not too many coding standard violations.
The code would look neater in packages.

Commit:
Duke A-MoreOOP: Use More OOP

Restructured the entire file to a more OOP structure. Java Doc is also updated.

Commit:
Duke A-Packages: Organize into Packages

Create 2 packages, Task and Storage. Subdivide classes into respective public class file.

Commit:
addition to A-Package

forgot to add Duke package

Commit:
Duke A-JUnit: Add JUnit Tests

Created two Junit Test to test the saving of file, and whether invalid inputs is not saved.

Commit:
Duke A-Jar: Create a JAR File

Created a workable Duke.java.jar file

Commit:
Level 9. Find

Give users a way to find a task by searching for a keyword.

Restructured ChoiceOfAction. Now it only chooses which Action Function to use.

New function doneAction, deleteAction and findAction.

Wilfredwongkc commented on src/main/java/Duke/DukeTest.java:
```diff
@@ -0,0 +1,56 @@
+
+public class DukeTest {
+    @Test
+    public void dummyTest(){
```
```suggestion
    public void dummyTest() {
```

Wilfredwongkc replied:
Maybe can add a space?

Wilfredwongkc commented on src/main/java/Storage/Storage.java:
```diff
@@ -0,0 +1,203 @@
+package Storage;
+
+import java.io.*;
```
Maybe can list out the things you want to import instead of using a wildcard?

Wilfredwongkc commented on src/main/java/Storage/Storage.java:
```diff
@@ -0,0 +1,203 @@
+import Task.EventsException;
+
+public class Storage {
+    private final Parser parser= new Parser();
```
```suggestion
    private final Parser PARSER = new Parser();
```

Wilfredwongkc replied:
I think if you are declaring a constant, it needs to be in caps. Also, don't forget the space before the "=".

Wilfredwongkc commented on src/main/java/Storage/Storage.java:
```diff
@@ -0,0 +1,203 @@
+
+// deals with making sense of the user command
+class Parser {
+
```
Don't forget to remove this empty line

Wilfredwongkc commented on src/main/java/Duke/DukeTest.java:
```diff
@@ -0,0 +1,56 @@
+
+    @Test
+    public void testInvalidFormat() {
+
```
Don't forget to remove this empty line too

Wilfredwongkc reviewed (changes requested):
LGTM! Just some nit that is related to coding stnadards to fix

Commit:
Duke A-Gradle: Use Gradle

Commit:
Duke Level-10: GUI

Changed many functions from TaskList and UI class such that they can return the message as String instead of printing it in the console. Setup the GUI which can print the results.

Commit:
new changes to level 10

mistakes were made, this is the actual level 10

Commit:
gradle issue

Commit:
Revert "gradle issue"

This reverts commit 8e28087fce190737fa8daf0d94c5f575f964961a.

Commit:
gradle issue

Commit:
jar message

1.Created a fat jar file underout\artifact\ip.jar.

Commit:
Created two assertions

Commit:
extension check duplicate

added new function isTaskDuplicate under parser.
The program can check if the input value will create a program that is already present in the tasklist

Commit:
C-DetectDuplicates

added new function isTaskDuplicate under parser.
The program can check if the input value will create a program that is already present in the tasklist

Commit:
Improve code quality

Improved code quality
1. changed multiple if-elseif statement to switch
2. Added new function comments
Remove a bug that causes program to print logo instead of parsing the input

Commit:
Improve code quality

Improved code quality
1. changed multiple if-elseif statement to switch
2. Added new function comments
Remove a bug that causes program to print logo instead of parsing the input

Commit:
try to merge

sj1999-BIT committed:
Merge branch 'master' into branch-A-CodeQuality

sj1999-BIT committed:
Merge pull request #4 from sj1999-BIT/branch-A-CodeQuality

Improve code quality

ramapriyan912001 commented on src/main/java/Storage/Storage.java:
```diff
@@ -0,0 +1,238 @@
+     * Function obtain data from previous file records. Creates the file if it's not present.
+     * @param lst is the target TaskList
+     */
+    public void getDataInputList(TaskList lst) {
+        try {
+            File f = new File("src/main/java/Duke.txt");
+            if(!f.exists()){
+                f.createNewFile();
+            }
+            FileReader reader = new FileReader(f);
+            BufferedReader br = new BufferedReader(reader);  //creates a buffering character input stream
+            String line;
+            Task curTask;
+
+            while ((line = br.readLine()) != null) {
+                boolean isCurTaskDone = false;
+                String[] data = line.split(" ");
+                int lengthLimit = data.length;
+                String taskData = "";
+
+                if (data[data.length - 1].equals("done")){
+                    isCurTaskDone = true;
+                    lengthLimit -= 1;
+                }
+                try {
+                    for (int i = 0; i < lengthLimit; i++) {
+                        taskData = taskData.concat(data[i] + " ");
+                    }
+                    curTask = parser.createTask(taskData);
+                    if (isCurTaskDone) {
+                        curTask.done();
+                    }
+                    lst.getTaskList().add(curTask);
+                } catch (Exception e) {
+                    System.out.println(e.getLocalizedMessage());
+                }
+            }
```
The method exceeds 40 lines.
Method could be SLAP-ed harder.

ramapriyan912001 commented on src/main/java/Storage/Storage.java:
```diff
@@ -0,0 +1,238 @@
+    }
+}
+
+class TodoTasks extends Task{
+    /**
+     * constructor for Task.
+     *
+     * @param action contains all the information to build a TodoTask.
+     * @throws TodoException if the String action is invalid.
+     */
+    public TodoTasks(String action) throws TodoException {
+        super(action, "[T][] ", TaskTypes.Todo);
+        if (action.split(" ").length < 2){
+            throw new TodoException("OOPS!!! The description of a todo cannot be empty.");
+        }
+    }
+}
+
+class DeadlineTask extends Task{
+    /**
+     * constructor for Task.
+     *
+     * @param action containinf all the data needed for the creation of Deadline Task.
+     * @throws DeadlineException if the String action is invalid.
+     */
+    public DeadlineTask(String action) throws DeadlineException {
+        super(action, "[D][] ", TaskTypes.Deadline);
+        if (action.split("/").length < 2) {
+            throw new DeadlineException("incorrect format for deadline task");
+        }
+        if (action.split("/")[1].split(" ").length < 2) {
+            throw new DeadlineException("incorrect date format for Deadline task");
+        }
+    }
+}
+
```
Should make these classes in separate java files rather than within another file.

ramapriyan912001 reviewed (commented):
Overall, the work is excellent. 
Some methods could be shortened perhaps.
Beyond that, great job!

Commit:
Improve the GUI and resolved bugs

Improved GUI by add different colors to Duke and User DialogBox.

Some error seemed to have occurred, reused previous local safed version to remove any errors.

Commit:
Merge commit '4d106abb4af3ac8469ad2384c142b5ab998ebdef'

sj1999-BIT committed:
Set theme jekyll-theme-minimal

sj1999-BIT committed:
Create index.md

Commit:
Merge branch 'master' of https://github.com/sj1999-BIT/ip

Commit:
Add Docs file

Docs file was previously deleted by mistake

Commit:
Create .gitignore

Commit:
stepping up user guide

Amended user guide, partially done.

Debugged code such that the bug whereby check duplicate function fails to work is resolved.

Commit:
remove lock files

trial

Commit:
Update README.md

Commit:
Update README.md

Commit:
Update README.md

Commit:
create a new jar file

smoke test done to test jar file validity

sj1999-BIT committed:
Update README.md

Improved syntax, removed the appearance of # in the UG

sj1999-BIT committed:
Update README.md

sj1999-BIT committed:
Update index.md
